criu/pie/restorer.c

```diff
@@ -1528,11 +1549,43 @@ long __export_restore_task(struct task_restore_args *args)
 	for (i = 0; i < args->vma_ios_n; i++) {
 		struct iovec *iovs = rio->iovs;
 		int nr = rio->nr_iovs;
-		ssize_t r;
+		ssize_t r = 0;
 
 		while (nr) {
-			pr_debug("Preadv %lx:%d... (%d iovs)\n", (unsigned long)iovs->iov_base, (int)iovs->iov_len, nr);
-			r = sys_preadv(args->vma_ios_fd, iovs, nr, rio->off);
+			/* Using mmap to load content as an on-demand way rather than egarly read all the content */
+			local_offset = rio->off;
+			for (j = 0; j < nr; j++) {
+				void * p = (void *)iovs[j].iov_base;
+				unsigned long end = (unsigned long)p + iovs[j].iov_len;
+				pr_debug("Change preadv to map of iov buffer %d, iov base %lx\n", j, (unsigned long)iovs[j].iov_base);
+				/* Considering the iov buffer may cross multiple regions */
+				while ((unsigned long)p < end) {
+					/* Find the vma_entry */
```

Comment:
too many indents usually means that you need to create a helper function

Reply:
I’ll deal with it later.

Reply:
let me know when this pr will be ready for review.

Reply:
Now it's ready for review.